Added dot product everywhere were cosine similarity was used

joancf opened the pull request:
## Related Issue

Support for inner product similarity measures

- #256

## Changes
almost everywhere (except some tests) where there was a cosine  reference a parallel dot function/option  is added 

What changed?

most files where cosine was used

## Testing and Validation

Not done , needs to be done, it  is pending. 
I'll try  to generate and replace the plugin in my ES isntallation and check it. 
but not sure which prodedures I must follow.

alexklibisz commented on elastiknn-api4s/src/test/scala/com/klibisz/elastiknn/api/XContentCodecSuite.scala:
```diff
@@ -110,13 +110,16 @@ class XContentCodecSuite extends AnyFreeSpec with Matchers {
           ("cosine", Similarity.Cosine),
           ("Cosine", Similarity.Cosine),
-          ("COSINE", Similarity.Cosine)
+          ("COSINE", Similarity.Cosine),
```
General comment for this file: Let's also add a block of tests below similar to this: https://github.com/alexklibisz/elastiknn/blob/39fa6109b3f03b5d03dbba404ead70b861fddceb/elastiknn-api4s/src/test/scala/com/klibisz/elastiknn/api/XContentCodecSuite.scala#L371-L406

alexklibisz commented on elastiknn-models/src/main/java/com/klibisz/elastiknn/models/DotLshModel.java:
```diff
@@ -0,0 +1,53 @@
+     * @param L number of hash tables
+     * @param k number of hash functions concatenated to form a hash for each table
+     * @param rng random number generator used to instantiate model parameters
+     */
```
```suggestion
    * @param vectorOps instance of FloatVectorOps used to execute vector operations.
    */
```

joancf commented:
Hi @alexklibisz 
First thanks for fast response, and the plug-in itself!! 
let me apologize , for sending the pull request before deeply testing it. It's my first time doing things in Scala, and I'm a bit confused on how to do some things.
Finally i could compile and build the zip in my side, so I can try to run the plugin!  (and check if it works and the performance) 
But as I said my knowledge of scala is my knowledge of Java. ... and for some of the things you ask me, i'm not able to do them. (basically testing is where I  think it will take a me a while to understand everything!)

About the changes you asked. I did all of them .
One I did in a different way was to ensure that the similarity was returning a positive value with max(0,1+dotProduct) 
In this way we don't raise an exception and negative values will have a 0 similartiy

Thanks
Joan

alexklibisz commented on docs/pages/api.md:
```diff
@@ -446,9 +470,12 @@ The exact transformations are described below.
 |Jaccard|N/A|0|1.0|
 |Hamming|N/A|0|1.0|
 |Cosine[^note-angular-cosine]|`cosine similarity + 1`|0|2|
+|Dot[^note-dot-product]|`Dot similarity + 1`|0|2| 
```
I think there might be something wrong with the table formatting: 
<img width="1598" alt="image" src="https://github.com/alexklibisz/elastiknn/assets/8015228/d44d0b53-cd09-4db0-b06f-8cdc7923d877">

Also, if we end up using it, we should describe the updated transformation here: `max(0, 1 + dot product)`

alexklibisz reviewed (commented):
Thanks for the quick fixes! I can help with the tests once we're happy with the implementation.

I'll have to think about more about the strategy for accounting for non-normalized vectors. As a user I think I would rather get a failed response than a bunch of missing results. Also, we say that the similarity is in [0, 2], but AFAICT the current strategy would allow for scores > 2, wouldn't it?

joancf commented on docs/pages/api.md:
```diff
@@ -446,9 +470,12 @@ The exact transformations are described below.
 |L1|`1 / (1 + l1 distance)`|0|1|
 |L2|`1 / (1 + l2 distance)`|0|1|
 
+Dot similirarity will produce negative scores if the vectors are not normalized
```
i added max(0,distance) so, it will always be positive.

alexklibisz commented:
Hi @joancf can you try adding the exceptions for scores outside [0, 2]? If you're having trouble I can try this, but probably not til the weekend or next week.

joancf commented:
hi. @alexklibisz my company doesn't want to use it. So, I will finish it out-hours 
I'll do my best with exceptions and testing.

joancf commented:
@alexklibisz  Now I'm back with this code, I wanted to extract some logs, 
 let me see If I can finish it

joancf commented:
Hi, @alexklibisz , 
I added some tests (i normalized the vectors for dot product) and also I rebased the branch to main
 I can't compile it, because there is an error in 
  [E008] Not Found Error: elastiknn-plugin/src/main/scala/com/klibisz/elastiknn/mapper/VectorMapper.scala:119:83 
[error] 119 |        new VectorMapper.FieldType(CONTENT_TYPE, context.buildFullName(thisBuilder.leafName), mapping),

maybe this a problem of the version of Elasticsearch I'm used ( I'm using 12.2 instead of 13)  but I need to work on that version and I don't want to upgrade my environment.

alexklibisz commented:
> Hi, @alexklibisz , I added some tests (i normalized the vectors for dot product) and also I rebased the branch to main I can't compile it, because there is an error in [E008] Not Found Error: elastiknn-plugin/src/main/scala/com/klibisz/elastiknn/mapper/VectorMapper.scala:119:83 [error] 119 | new VectorMapper.FieldType(CONTENT_TYPE, context.buildFullName(thisBuilder.leafName), mapping),
> 
> maybe this a problem of the version of Elasticsearch I'm used ( I'm using 12.2 instead of 13) but I need to work on that version and I don't want to upgrade my environment.

In order to get this merged, it needs to be working with the latest version. If that doesn't work, you're welcome to fork and build your own artifacts for older versions. I'm not maintaining multiple/older versions of Elasticsearch.

joancf commented:
ok, 
Alex I synchronized  my repo, Now I can compile and generate the plugin 8.15.0.0
